### PR TITLE
TT-511 containerize time tracker UI

### PR DESCRIPTION
**Pull request by mikevillarruel**

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -0,0 +1,50 @@
+.PHONY: help
+help:
+	@echo "---------------HELP-----------------"
+	@echo "- make build --> Create docker image with dependencies needed"
+	@echo "- make run --> Execute docker container database from postgres, and api from image created previusly"
```
Comment (OscarPallazhco) [inline Makefile:5]: you are not executing a container database because your docker-compose.yml file only has the time-tracker-ui service.
```suggestion
	@echo "- make run --> Execute timetracker_ui docker container."
```

File: Dockerfile [hunk 2]
```diff
@@ -0,0 +1,7 @@
+FROM node:14
+
+WORKDIR /time-tracker-ui
+
+COPY package*.json .
```
Comment (OscarPallazhco) [inline Dockerfile:5]: I think it is not necessary, because you are using COPY . .

**Review by OscarPallazhco — APPROVED** [review]

**Review by MichaelGuanoluisa — APPROVED** [review]

**Review by MichaelAndrik — APPROVED** [review]

File: Dockerfile [hunk 3]
```diff
@@ -0,0 +1,45 @@
+FROM node:14 AS development
+
+ENV USERNAME timetracker
+ENV HOME /home/${USERNAME}
+
+RUN useradd -ms /bin/bash ${USERNAME}
+
+WORKDIR ${HOME}/time-tracker-ui
+COPY [^.env]* . .
+RUN chown ${USERNAME}:${USERNAME} -R ${HOME}
```
Comment (faustocv) [inline Dockerfile:10]: Why is this needed? the home folder already belongs to the new user "timetracker"

File: Dockerfile [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+RUN npm install
+RUN npm run build
+
+
+
+FROM nginx:1.21 AS production
+
+ENV USERNAME nginx
+
+COPY nginx.conf /etc/nginx/conf.d/
```
Comment (faustocv) [inline Dockerfile:34]: Why is this needed? because the default.conf one is being copied after.

File: Dockerfile [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+COPY ./ /usr/local/app/
+
+RUN npm install
+RUN npm run build
+
+
+
+FROM nginx:1.21 AS production
+
+ENV USERNAME nginx
```
Comment (faustocv) [inline Dockerfile:32]: Please, instead of nginx name it as app

File: Dockerfile [hunk 6]
```diff
@@ -0,0 +1,45 @@
+FROM node:14 AS development
```
Comment (faustocv) [inline Dockerfile:1]: I believe this might be the base stage and then fellow stages could be start from it

File: Makefile [hunk 7]
```diff
@@ -0,0 +1,77 @@
+.PHONY: help
+help:
```
Comment (faustocv) [inline Makefile:2]: Please for this help task follow how was done at https://github.com/ioet/infra-terraform-modules/blob/fb270d25da76abacb4ce6fdac8ac3e3e8ef05a57/Makefile#L11

File: Makefile [hunk 8]
```diff
@@ -0,0 +1,77 @@
...
+	@echo "- make run_prod --> Execute timetracker_ui_prod docker container"
+	@echo "- make remove_prod --> Delete container timetracker_ui_prod"
+	@echo "- make publish_prod --> Publish the container image timetracker_ui_prod"
+
+	@echo "- make login --> Login in respository of docker images"
+	@echo "------------------------------------"
+
+.PHONY: build
+build:
+	-docker rmi timetracker_ui
```
Comment (faustocv) [inline Makefile:23]: Extract rmi to another make task named as `cleanup`

File: Makefile [hunk 9]
```diff
@@ -0,0 +1,77 @@
...
+	docker-compose build
+
+.PHONY: run
+run:
+	docker-compose --env-file ./.env up -d
+	docker logs -f timetracker_ui
```
Comment (faustocv) [inline Makefile:29]: Extract `docker logs` to another make task named as `logs`

File: Makefile [hunk 10]
```diff
@@ -0,0 +1,77 @@
...
+
+.PHONY: stop
+stop:
+	docker-compose stop	
+
+.PHONY: restart
+restart:
+	docker-compose stop	
+	docker-compose up -d
+	docker logs -f timetracker_ui
```
Comment (faustocv) [inline Makefile:39]: Remove this

File: Makefile [hunk 11]
```diff
@@ -0,0 +1,77 @@
...
+	docker-compose down
+
+.PHONY: test
+test:
+	docker-compose --env-file ./.env up -d
+	docker exec -it timetracker_ui bash -c "npm run test"
+
+.PHONY: publish
+publish:
+	docker tag timetracker_ui:latest timetrackerdevregistry.azurecr.io/timetracker_ui:latest
```
Comment (faustocv) [inline Makefile:52]: Parametrize registry url

File: Makefile [hunk 12]
```diff
@@ -0,0 +1,77 @@
...
+	docker push timetrackerdevregistry.azurecr.io/timetracker_ui:latest
+
+.PHONY: build_prod
+build_prod:
+	-docker rmi timetracker_ui_prod
```
Comment (faustocv) [inline Makefile:57]: Remove this

File: Makefile [hunk 13]
```diff
@@ -0,0 +1,77 @@
...
+	docker build --target production -t timetracker_ui_prod -f Dockerfile .
+
+.PHONY: run_prod
+run_prod:
+	docker run -d -p 4200:4200 --name timetracker_ui_prod timetracker_ui_prod 
+	docker logs -f timetracker_ui_prod
```
Comment (faustocv) [inline Makefile:63]: Remove this

File: Makefile [hunk 14]
```diff
@@ -0,0 +1,77 @@
...
+
+.PHONY: remove_prod
+remove_prod:
+	docker stop timetracker_ui_prod
+	docker rm timetracker_ui_prod
+
+.PHONY: publish_prod
+publish_prod:
+	docker tag timetracker_ui_prod:latest timetrackerdevregistry.azurecr.io/timetracker_ui_prod:latest
```
Comment (faustocv) [inline Makefile:72]: Parametrize registry url

File: Dockerfile (hunk 3 above)
Comment (mikevillarruel) [reply Dockerfile]: Yes home folder belongs to the new user, but when I copy files inside of that folder this belongs to the root instead of the new user, for that reason I have some permission issues when I try to execute the project in the container and also when I try to install npm packages, that's why I need do that. Currently I change this to give permissions to the work directory for be more specific, because home actually belongs to the new user.

File: Dockerfile (hunk 4 above)
Comment (mikevillarruel) [reply Dockerfile]: You're right, I removed that line.

File: Dockerfile (hunk 5 above)
Comment (mikevillarruel) [reply Dockerfile]: Done

File: Dockerfile (hunk 6 above)
Comment (mikevillarruel) [reply Dockerfile:1]: Done, currently this is the base image to the build stage

File: Makefile (hunk 7 above)
Comment (mikevillarruel) [reply Makefile]: Done

File: Makefile (hunk 10 above)
Comment (mikevillarruel) [reply Makefile]: Done

File: Makefile (hunk 11 above)
Comment (mikevillarruel) [reply Makefile]: Done

File: Makefile (hunk 12 above)
Comment (mikevillarruel) [reply Makefile]: Done

File: Makefile (hunk 13 above)
Comment (mikevillarruel) [reply Makefile]: Done

File: Makefile (hunk 14 above)
Comment (mikevillarruel) [reply Makefile]: Done

File: Makefile (hunk 9 above)
Comment (mikevillarruel) [reply Makefile:26]: Done

File: Makefile (hunk 8 above)
Comment (mikevillarruel) [reply Makefile]: Done

File: Makefile [hunk 15]
```diff
@@ -0,0 +1,71 @@
...
+	docker-compose stop	
+
+.PHONY: restart
+restart: ## Restart container timetracker_ui.
+	docker-compose stop	
+	docker-compose up -d
+
+.PHONY: remove
+remove: ## Delete container timetracker_ui.
+	docker-compose down
```
Comment (faustocv) [inline Makefile:39]: Add `docker-compose down --volumes --remove-orphans --rmi local`

File: Makefile [hunk 16]
```diff
@@ -0,0 +1,71 @@
...
+	docker rm timetracker_ui_prod
+
+.PHONY: publish_prod
+publish_prod: ## Publish the container image timetracker_ui_prod.
+	docker tag timetracker_ui_prod:latest $(registry_url)/timetracker_ui_prod:latest
+	docker push $(registry_url)/timetracker_ui_prod:latest
+
+.PHONY: login
+login: ## Login in respository of docker images.
+	az acr login --name timetrackerdevregistry
```
Comment (faustocv) [inline Makefile:71]: Could you allow specify the container registry from a parameter?

File: Makefile (hunk 15 above)
Comment (mikevillarruel) [reply Makefile]: Done

File: Makefile (hunk 16 above)
Comment (mikevillarruel) [reply Makefile]: Done

**Review by faustocv — APPROVED** [review]
Awesome. Thank you for doing this
